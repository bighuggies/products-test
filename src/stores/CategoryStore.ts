import { action, computed, decorate, observable } from 'mobx';

import { ICategory } from '../models/Category';

export class CategoryStore {
  public static STORE_NAME = 'categoryStore';

  constructor(
    public categories: ICategory[] = [],
    public currentCategoryId?: string
  ) {}

  public fetchCategories = async () => {
    const response = await fetch(
      'https://api.gousto.co.uk/products/v2.0/categories'
    );

    const data = await response.json();

    this.setCategories(data.data);
  };

  public setCategories = (categories: ICategory[]) => {
    this.categories = categories;
  };

  public get activeCategories() {
    return this.categories.filter(c => !c.hidden);
  }

  public get currentCategory() {
    return this.categories.find(c => c.id === this.currentCategoryId);
  }
}

decorate(CategoryStore, {
  activeCategories: computed,
  categories: observable,
  currentCategory: computed,
  currentCategoryId: observable,
  setCategories: action
});
