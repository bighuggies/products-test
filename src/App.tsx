import { inject, observer } from 'mobx-react';
import * as React from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';

import './App.css';
import Category from './components/Category';
import CategoryMenu from './components/CategoryMenu';
import { CategoryStore } from './stores/CategoryStore';

class App extends React.Component<{ categoryStore?: CategoryStore }> {
  public componentDidMount() {
    if (
      this.props.categoryStore &&
      this.props.categoryStore.categories.length <= 0
    ) {
      this.props.categoryStore.fetchCategories();
    }
  }

  public render() {
    return (
      this.props.categoryStore && (
        <Router>
          <div>
            <CategoryMenu />

            <Route path="/:categoryId" component={Category} />
          </div>
        </Router>
      )
    );
  }
}

export default inject(CategoryStore.STORE_NAME)(observer(App));
