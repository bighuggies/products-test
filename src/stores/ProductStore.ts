import { action, computed, decorate, observable } from 'mobx';

import { IProduct } from '../models/Product';

import { CategoryStore } from './CategoryStore';

export class ProductStore {
  public static STORE_NAME = 'productStore';

  constructor(
    public categoryStore: CategoryStore,
    public products: IProduct[] = [],
    public filter: string = ''
  ) {}

  public fetchProducts = async () => {
    const response = await fetch(
      'https://api.gousto.co.uk/products/v2.0/products?includes[]=categories&includes[]=attributes&sort=position&image_sizes[]=365&image_sizes[]=400&period_id=120'
    );
    const data = await response.json();

    this.setProducts(data.data);
  };

  public selectCategory = (categoryId: string) =>
    this.categoryStore.selectCategory(categoryId);

  public setProducts = (products: IProduct[]) => (this.products = products);

  public get categoryProducts(): IProduct[] {
    return this.categoryStore.currentCategoryId && this.products
      ? this.products.filter(p =>
          p.categories.some(c => c.id === this.categoryStore.currentCategoryId)
        )
      : [];
  }

  public get filteredCategoryProducts(): IProduct[] {
    return this.categoryProducts.filter(
      c =>
        !this.filter ||
        c.title.toLowerCase().includes(this.filter.toLowerCase()) ||
        c.description.toLowerCase().includes(this.filter.toLowerCase())
    );
  }
}

decorate(ProductStore, {
  categoryProducts: computed,
  filter: observable,
  filteredCategoryProducts: computed,
  products: observable,
  setProducts: action
});
