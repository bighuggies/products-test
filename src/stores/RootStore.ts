import { CategoryStore } from './CategoryStore';
import { ProductStore } from './ProductStore';

export class RootStore {
  constructor() {
    this[ProductStore.STORE_NAME] = new ProductStore();
    this[CategoryStore.STORE_NAME] = new CategoryStore();
  }
}
