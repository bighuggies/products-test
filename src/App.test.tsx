import { mount } from 'enzyme';
import { Provider } from 'mobx-react';
import * as React from 'react';

import App from './App';
import { RootStore } from './stores/RootStore';

test('renders without crashing', () => {
  const mounted = mount(
    <Provider {...new RootStore()}>
      <App />
    </Provider>
  );

  expect(mounted.debug()).toMatchSnapshot();
});
