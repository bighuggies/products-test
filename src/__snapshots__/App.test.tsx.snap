// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
"<Provider categoryStore={{...}} productStore={{...}}>
  <inject-App-with-categoryStore>
    <App categoryStore={{...}}>
      <BrowserRouter>
        <Router history={{...}}>
          <div>
            <inject-withRouter(CategoryMenu)-with-categoryStore>
              <withRouter(CategoryMenu) categoryStore={{...}}>
                <Route>
                  <CategoryMenu categoryStore={{...}} match={{...}} location={{...}} history={{...}} staticContext={[undefined]}>
                    <styled.ul>
                      <ul className=\\"sc-bdVaJa hCdHre\\" />
                    </styled.ul>
                  </CategoryMenu>
                </Route>
              </withRouter(CategoryMenu)>
            </inject-withRouter(CategoryMenu)-with-categoryStore>
            <Route path=\\"/:categoryId\\" component={[Function: Injector]} />
          </div>
        </Router>
      </BrowserRouter>
    </App>
  </inject-App-with-categoryStore>
</Provider>"
`;
